libraries/c_sdk/standard/ble/test/iot_mqtt_ble_system_test.c

```diff
@@ -0,0 +1,1040 @@
+ * @param[in] pPacketInfo Info for the incoming acknowledgement packet.
+ * @param[in] packetIdentifier The ID of the incoming packet.
+ */
+static void handleAckEvents( MQTTPacketInfo_t * pPacketInfo,
+                             uint16_t packetIdentifier );
+
+/**
+ * @brief The application callback function that is expected to be invoked by the
+ * MQTT library for incoming publish and incoming acks received over the network.
+ *
+ * @param[in] pContext MQTT context pointer.
+ * @param[in] pPacketInfo Packet Info pointer for the incoming packet.
+ * @param[in] pDeserializedInfo Deserialized information from the incoming packet.
+ */
+static void eventCallback( MQTTContext_t * pContext,
+                           MQTTPacketInfo_t * pPacketInfo,
+                           MQTTDeserializedInfo_t * pDeserializedInfo );
+
+/**
+ * The function for initializing the BLE stack.
+ * The function implementation is provided by the vendor.
+ */
+extern BTStatus_t bleStackInit( void );
+
+/*
+ * Callback invoked when physical BLE connection is established.
+ */
+static void bleConnectionCallback( BTStatus_t status,
+                                   uint16_t connId,
+                                   bool connected,
+                                   BTBdaddr_t * pBa );
+
+/**
+ * Enable BLE and GATT services.
+ * Waits for the BLE phyical connection to be established.
+ */
+static void bleEnable( void );
+
+/**
+ * Disconnect and turn off BLE.
+ */
+static void bleDisable( void );
+
+/**
+ * Sets up transport interface used for sending and receiving MQTT data.
+ */
+static void setupBleTransportInterface( NetworkContext_t * pContext );
+
+/*
+ * Cleans up transport interface used for sending receiving MQTT data.
+ */
+static void teardownBleTransportInterface( NetworkContext_t * pContext );
+
+/*-----------------------------------------------------------*/
+
+static void _BLEConnectionCallback( BTStatus_t status,
+                                    uint16_t connId,
+                                    bool connected,
+                                    BTBdaddr_t * pBa )
+{
+    if( connected == true )
+    {
+        IotBle_StopAdv( NULL );
+    }
+    else
+    {
+        ( void ) IotBle_StartAdv( NULL );
+    }
+
+    isBleConnected = connected;
+    IotSemaphore_Post( &bleConnectionSem );
+}
+
+
+static void bleEnable( void )
+{
+    IotBleEventsCallbacks_t eventCb;
+    BTStatus_t status = eBTStatusSuccess;
+    static bool isBleStackInitialized = false;
+
+    /** One time initialization of stack and Middleware **/
+    if( isBleStackInitialized == false )
+    {
+        status = bleStackInit();
+
+        TEST_ASSERT_EQUAL( eBTStatusSuccess, status );
+
+        status = IotBle_Init();
+
+        TEST_ASSERT_EQUAL( eBTStatusSuccess, status );
+
+        isBleStackInitialized = true;
+    }
+
+    TEST_ASSERT_EQUAL_INT( true, IotSemaphore_Create( &bleConnectionSem, 0, 1 ) );
```

Comment:
Same comment about the abstraction for Semaphore and all the instances in this file

Reply:
Fixed.